text_generation/causal_lm/cpp/prompt_lookup_decoding.cpp

```diff
@@ -0,0 +1,308 @@
+        std::string text = detokenize(detokenizer, token_cache);
+        std::cout << std::string_view{text.data() + print_len, text.size() - print_len} << '\n';
+        token_cache.clear();
+        print_len = 0;
+    }
+};
+
+ov::Tensor trimm_tensor(ov::Tensor& tensor, uint64_t seq_len_axis, uint64_t new_seq_len) {
+    // Copy elements from the old to a new tensor and return it.
+    // It's assumed that key/values tensor has a shape [BATCH_SIZE, num_kv_heads, seq_len, head_size] or [seq_len, ...],
+    // It that's not the case for your model please implement your own trim method.
+    OPENVINO_ASSERT(seq_len_axis == 2 || seq_len_axis == 0,
+                    "Cannot trim key/values with sequence length axis = ",
+                    seq_len_axis);
+
+    auto old_tensor_data = tensor.data<float>();
+    auto shape = tensor.get_shape();
+    size_t batch_size = shape[0];
+    size_t num_kv_heads = shape[1];
+    size_t old_seq_len = shape[2];
+    size_t head_size = shape[3];
+
+    OPENVINO_ASSERT(new_seq_len <= old_seq_len);
+
+    // if new_seq_len equal to old one no need to copy tensor, return as is
+    if (old_seq_len == new_seq_len)
+        return tensor;
+
+    if (seq_len_axis == 0) {
+        shape[0] = new_seq_len;
+        tensor.set_shape(shape);
+    }
+
+    // if seq_len_axis == 2, then data is not contiguous, in order to trim need to repack tensor
+    auto new_tensor = ov::Tensor{ov::element::f32, {batch_size, num_kv_heads, new_seq_len, head_size}};
+    auto new_tensor_data = new_tensor.data<float>();
+    for (size_t batch = 0; batch < batch_size; ++batch) {
+        for (size_t i = 0; i < num_kv_heads; ++i) {
+            for (size_t j = 0; j < new_seq_len; ++j) {
+                auto dst_ptr = new_tensor_data + num_kv_heads * new_seq_len * head_size * batch +
+                               new_seq_len * head_size * i + head_size * j;
+                auto src_ptr = old_tensor_data + num_kv_heads * new_seq_len * head_size * batch +
+                               old_seq_len * head_size * i + head_size * j;
+                std::memcpy(dst_ptr, src_ptr, head_size * sizeof(float));
+            }
+        }
+    }
+    return new_tensor;
+}
+
+void update_kv_cache(ov::InferRequest request, uint64_t seq_len_axis, uint64_t new_seq_len) {
+    // trim kv_cache values up to the new_seq_len
+    for (auto& state : request.query_state()) {
+        ov::Tensor old_tensor = state.get_state();
+        state.set_state(trimm_tensor(old_tensor, seq_len_axis, new_seq_len));
+    }
+}
+
+class PromptLookupCandidateGenerator {
+private:
+    const size_t max_ngram_size = 3;
+    size_t num_pred_tokens = 5;
+    const size_t max_pred_tokens = 20;
+
+public:
+    PromptLookupCandidateGenerator(const size_t max_ngram_size, const size_t num_pred_tokens)
+        : max_ngram_size{max_ngram_size},
+          num_pred_tokens{num_pred_tokens} {};
+
+    std::vector<int64_t> generate_candidates(const std::vector<int64_t>& input_ids) {
```

Comment:
I just want to verify. `for` is a common 3 symbol sequence. It's likely to be encountered multiple times in the prompt. `generate_candidates()` always returns the sequence which follows the first `for` occurrence. For the next `for` the returned candidates won't be useful. Is there anything to fix that?